extendeddatacrossword.go

```diff
@@ -94,87 +109,128 @@ func (eds *ExtendedDataSquare) solveCrossword(rowRoots [][]byte, columnRoots [][
 		for i := 0; i < int(eds.width); i++ {
 			for mode := range []int{row, column} {
-
 				var isIncomplete bool
 				var isExtendedPartIncomplete bool
-				if mode == row {
+				switch mode {
+				case row:
 					isIncomplete = !bitMask.RowIsOne(i)
 					isExtendedPartIncomplete = !bitMask.RowRangeIsOne(i, int(eds.originalDataWidth), int(eds.width))
-				} else if mode == column {
+				case column:
 					isIncomplete = !bitMask.ColumnIsOne(i)
 					isExtendedPartIncomplete = !bitMask.ColRangeIsOne(i, int(eds.originalDataWidth), int(eds.width))
+				default:
+					panic(fmt.Sprintf("invalid mode %d", mode))
 				}
 
 				if isIncomplete { // row/column incomplete
 					// Prepare shares
 					shares := make([][]byte, eds.width)
 					for j := 0; j < int(eds.width); j++ {
 						var vectorData [][]byte
-						var rowIdx, colIdx int
-						if mode == row {
-							rowIdx = i
-							colIdx = j
+						var r, c int
+						switch mode {
+						case row:
+							r = i
+							c = j
 							vectorData = eds.Row(uint(i))
-						} else if mode == column {
-							rowIdx = j
-							colIdx = i
+						case column:
+							r = j
+							c = i
 							vectorData = eds.Column(uint(i))
+						default:
+							panic(fmt.Sprintf("invalid mode %d", mode))
 						}
-						if bitMask.Get(rowIdx, colIdx) {
+						if bitMask.Get(r, c) {
 							// As guaranteed by the bitMask, vectorData can't be nil here:
 							shares[j] = vectorData[j]
 						}
 					}
 
 					// Attempt rebuild
 					rebuiltShares, err := Decode(shares, eds.codec)
-					if err != nil { // repair unsuccessful
+					if err != nil {
+						// repair unsuccessful
 						solved = false
-					} else { // repair successful
-						progressMade = true
-						// Insert rebuilt shares into square
-						for p, s := range rebuiltShares {
-							if mode == row {
-								eds.setCell(uint(i), uint(p), s)
-							} else if mode == column {
-								eds.setCell(uint(p), uint(i), s)
-							}
-						}
-						if isExtendedPartIncomplete {
-							err := eds.rebuildExtendedPart(mode, uint(i))
-							if err != nil {
-								return err
-							}
-						}
+						continue
+					}
 
-						// Check that rebuilt vector matches given merkle root
-						err := eds.verifyRoots(rowRoots, columnRoots, mode, uint(i))
+					progressMade = true
+
+					if isExtendedPartIncomplete {
+						// If needed, rebuild the parity shares too.
+						rebuiltExtendedShares, err := Encode(rebuiltShares[0:eds.originalDataWidth], eds.codec)
 						if err != nil {
 							return err
 						}
+						startIndex := len(rebuiltExtendedShares) - int(eds.originalDataWidth)
+						rebuiltShares = append(
+							rebuiltShares[0:eds.originalDataWidth],
+							rebuiltExtendedShares[startIndex:]...,
+						)
+					} else {
+						// Otherwise copy them from the EDS.
+						startIndex := len(shares) - int(eds.originalDataWidth)
+						rebuiltShares = append(
+							rebuiltShares[0:eds.originalDataWidth],
```

Comment:
Can you still open an issue in the go-leopard library about the diff in the API/returned results.

Reply:
Opened https://github.com/lazyledger/go-leopard/issues/7